presto-druid/src/main/java/com/facebook/presto/druid/DruidSplitManager.java

```diff
@@ -52,16 +69,119 @@ public ConnectorSplitSource getSplits(
+                DruidColumnHandle handle = (DruidColumnHandle) entry.getKey();
+                if (handle.getColumnName().equals("__time")) {
+                    Domain predicate = entry.getValue();
+                    segmentFilter = segmentFilter(predicate);
+                }
+                else {
+                    DruidColumnType dct = DruidColumnType.fromPrestoType(handle.getColumnType());
+                    if (dct == DruidColumnType.VARCHAR) {
+                        List<Object> values = stringDimensions(entry.getValue());
+                        if (values != null && !values.isEmpty()) {
+                            dimFilters.put(handle.getColumnName(), values);
+                        }
+                    }
+                }
+            }
+            if (segmentFilter != null) {
+                segmentIds = druidClient.getDataSegmentId(table.getTableName(), segmentFilter);
+            }
+            else {
+                segmentIds = druidClient.getDataSegmentId(table.getTableName());
+            }
+        }
+        else {
+            segmentIds = druidClient.getDataSegmentId(table.getTableName());
+        }
         List<DruidSplit> splits = segmentIds.stream()
                 .map(id -> druidClient.getSingleSegmentInfo(table.getTableName(), id))
-                .map(info -> createSegmentSplit(info, HostAddress.fromUri(druidClient.getDruidBroker())))
+                .map(info -> createSegmentSplit(info, HostAddress.fromUri(druidClient.getDruidBroker()), dimFilters))
                 .collect(toImmutableList());
 
         return new FixedSplitSource(splits);
     }
+
+    private List<Object> stringDimensions(Domain predicate)
+    {
+        List<Object> dimensions = new ArrayList<>();
+        ValueSet values = predicate.getValues();
+        if (values instanceof SortedRangeSet) {
+            SortedRangeSet srs = (SortedRangeSet) values;
+            List<Range> ranges = srs.getOrderedRanges();
+            for (Range range : ranges) {
+                if (range.getLow().getValueBlock().isPresent() && range.getLow().getBound() == Bound.EXACTLY) {
+                    dimensions.add(((Slice) range.getLow().getValue()).toStringUtf8());
+                }
+            }
+        }
+        return dimensions;
+    }
+
+    private String segmentFilter(Domain predicate)
+    {
+        ValueSet values = predicate.getValues();
+        if (values instanceof SortedRangeSet) {
+            SortedRangeSet srs = (SortedRangeSet) values;
+            List<Range> ranges = srs.getOrderedRanges();
+            if (ranges.size() > 1) {
+                // In clause style predicate, we can't use IN because druid stores it as text.
+                List<String> filters = new ArrayList<>();
+                for (Range range : ranges) {
+                    String iso8601 = timeSegmentFormatter.format(Instant.ofEpochMilli((Long) range.getSingleValue()));
```

Comment:
I guess we can have a better name here,  isoDateTime, isoTimestamp, iso8601DateString or something?  your call

Reply:
agree, I'll name it better.